### PR TITLE
Remove unused/unreachable code from ops

### PR DESCRIPTION
**Pull request by jbrockmendel**
Add TODO comments for questionable cases.

- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/core/ops.py [hunk 1]
```diff
@@ -721,6 +686,10 @@ def safe_na_op(lvalues, rvalues):
                 return na_op(lvalues, rvalues)
         except Exception:
             if isinstance(rvalues, ABCSeries):
+                # TODO: This case is not hit in tests.  For it to be hit would
+                # require `right` to be an object such that right.values
```
Comment (jreback) [inline pandas/core/ops.py]: limit the comments to useful info, 'screwup' does not fall in this category

File: pandas/core/ops.py [hunk 2]
```diff
@@ -731,7 +700,9 @@ def safe_na_op(lvalues, rvalues):
                                                   lambda x: op(x, rvalues))
             raise
 
-    def wrapper(left, right, name=name, na_op=na_op):
+    def wrapper(left, right, na_op=na_op):
+        # TODO: `na_op` does not belong in Series.__{op}__ signature.  But this
```
Comment (jreback) [inline pandas/core/ops.py]: huh?

File: pandas/core/ops.py [hunk 3]
```diff
@@ -740,32 +711,27 @@ def wrapper(left, right, name=name, na_op=na_op):
...
-            left,
-            result,
-            index=left.index,
-            name=name,
-            dtype=dtype,
-        )
+        return construct_result(left, result,
+                                index=left.index, name=res_name, dtype=dtype)
 
+    wrapper.__name__ = name
```
Comment (jreback) [inline pandas/core/ops.py]: I believe this is hit in tests, yes? (the name of the wrapper)

File: pandas/core/ops.py [hunk 4]
```diff
@@ -731,7 +700,9 @@ def safe_na_op(lvalues, rvalues):
                                                   lambda x: op(x, rvalues))
             raise
 
-    def wrapper(left, right, name=name, na_op=na_op):
+    def wrapper(left, right, na_op=na_op):
+        # TODO: `na_op` does not belong in Series.__{op}__ signature.  But this
+        # is needed here for closure/namespace purposes ATM.
```
Comment (jreback) [inline pandas/core/ops.py]: can you add doc-strings. this is extremely useful to future readers.

File: pandas/core/ops.py [hunk 5]
```diff
@@ -1371,23 +1339,6 @@ def f(self, other):
 
 def _arith_method_PANEL(op, name, str_rep=None, fill_zeros=None,
                         default_axis=None, **eval_kwargs):
-    # copied from Series na_op above, but without unnecessary branch for
-    # non-scalar
```
Comment (jreback) [inline pandas/core/ops.py]: ok, not super concerned about Panel things generally.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops.py [hunk 6]
```diff
@@ -440,44 +442,11 @@ def _validate_timedelta(self, name):
...
-            raise TypeError('cannot operate on a series without a rhs '
-                            'of a series/ndarray of type datetime64[ns] '
-                            'or a timedelta')
-
     def _validate(self, lvalues, rvalues, name):
         if self.is_datetime_lhs:
             return self._validate_datetime(lvalues, rvalues, name)
         elif self.is_timedelta_lhs:
             return self._validate_timedelta(name)
-        elif self.is_offset_lhs:
```
Comment (jreback) [inline pandas/core/ops.py]: why is this offset case not possible? (it may not be tested) but is it possible?

Comment (jbrockmendel) [reply pandas/core/ops.py]: Because _TimeOp is only used if `is_datetime_lhs or is_timedelta_lhs`

File: pandas/core/ops.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: Right now:
```
>>> pd.Series.__sub__
<unbound method Series.wrapper>
```

File: pandas/core/ops.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: 0.21.1:
```
>>> inspect.getargspec(pd.Series.__add__)
ArgSpec(args=['left', 'right', 'name', 'na_op'], varargs=None, keywords=None, defaults=('__add__', <function na_op at 0x10c4b7d70>))
```

File: pandas/core/ops.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: This comment is more a suggestion that we get rid of this block.

File: pandas/core/ops.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: Any docstring added here would become the docstring for `pd.Series.__foo__`, which I doubt is what you have in mind here.

File: pandas/core/ops.py (hunk 6 above)
Comment (jreback) [reply pandas/core/ops.py]: ok, can you doc-string things a bit, will help future readers.

File: pandas/core/ops.py (hunk 3 above)
Comment (jreback) [reply pandas/core/ops.py]: hmm yeah that's not right. can you add tests for this? (I believe we have a suite of tests for the stat methods, so prob just need to add these on).

File: pandas/core/ops.py (hunk 2 above)
Comment (jreback) [reply pandas/core/ops.py]: and so what?

File: pandas/core/ops.py (hunk 1 above)
Comment (jreback) [reply pandas/core/ops.py]: I get that. but this is public code, you can certainly say 'this should be removed'

File: pandas/core/ops.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: We recently made similar tests for Index comparison method names in tests.indexes.test_base, I don't see much else to that effect.  I'll put something together.

File: pandas/core/ops.py (hunk 4 above)
Comment (jreback) [reply pandas/core/ops.py]: right, so just a comment then which describes the input & types

File: pandas/core/ops.py (hunk 3 above)
Comment (jreback) [reply pandas/core/ops.py]: https://github.com/pandas-dev/pandas/blob/master/pandas/tests/generic/test_generic.py#L507

File: pandas/core/ops.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: ... and there's [precedent](https://github.com/pandas-dev/pandas/pull/17179#discussion_r132919572) for caring about signatures that make sense, especially in super-commonly-used user-facing methods.

Comment (jbrockmendel) [reply pandas/core/ops.py]: Don't worry about it anyway, I found a way to get na_op and name out of the signature.

File: pandas/core/ops.py [hunk 7]
```diff
@@ -745,35 +718,31 @@ def wrapper(left, right, name=name, na_op=na_op):
...
-            index=left.index,
-            name=name,
-            dtype=dtype,
-        )
+        result = wrap_results(safe_na_op(lvalues, rvalues, converted.na_op))
+        return construct_result(left, result,
+                                index=left.index, name=res_name, dtype=dtype)
 
+    wrapper.__name__ = name
+    if hasattr(operator, name) and callable(getattr(operator, name)):
```
Comment (jreback) [inline pandas/core/ops.py]: huh? what is this for

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: You asked for generic docstrings for these ops.  These are about as generic as they get.  But there is no e.g. operator.rfloordiv, so we have to condition.

File: pandas/core/ops.py [hunk 8]
```diff
@@ -354,24 +356,25 @@ class _TimeOp(_Op):
     """
     Wrapper around Series datetime/time/timedelta arithmetic operations.
     Generally, you should use classmethod ``_Op.get_op`` as an entry point.
+
+    This is only reached in cases where either `self.is_datetime_lhs`
```
Comment (jreback) [inline pandas/core/ops.py]: you don't need the single back ticks

File: pandas/core/ops.py [hunk 9]
```diff
@@ -354,24 +356,25 @@ class _TimeOp(_Op):
...
         # left
-        self.is_offset_lhs = is_offsetlike(left)
         self.is_timedelta_lhs = is_timedelta64_dtype(lvalues)
         self.is_datetime64_lhs = is_datetime64_dtype(lvalues)
         self.is_datetime64tz_lhs = is_datetime64tz_dtype(lvalues)
         self.is_datetime_lhs = (self.is_datetime64_lhs or
                                 self.is_datetime64tz_lhs)
-        self.is_integer_lhs = left.dtype.kind in ['i', 'u']
-        self.is_floating_lhs = left.dtype.kind == 'f'
+        assert left.dtype.kind not in ['i', 'u', 'f'], left
```
Comment (jreback) [inline pandas/core/ops.py]: assert one of the is_* e.g.
``asert self.is_timedelta_lhs or self is_datetime_lhs``

File: pandas/core/ops.py [hunk 10]
```diff
@@ -517,15 +489,16 @@ def _convert_to_array(self, values, name=None, other=None):
             elif isinstance(values, pd.DatetimeIndex):
                 values = values.to_series()
             # datetime with tz
```
Comment (jreback) [inline pandas/core/ops.py]: fix this comment

File: pandas/core/ops.py [hunk 11]
```diff
@@ -517,15 +489,16 @@ def _convert_to_array(self, values, name=None, other=None):
...
-                  hasattr(ovalues, 'tzinfo')):
+            elif isinstance(ovalues, datetime.datetime):
                 values = pd.DatetimeIndex(values)
             # datetime array with tz
             elif is_datetimetz(values):
                 if isinstance(values, ABCSeries):
                     values = values._values
             elif not (isinstance(values, (np.ndarray, ABCSeries)) and
                       is_datetime64_dtype(values)):
+                # TODO: This is not hit in tests.  What case is it intended
```
Comment (jreback) [inline pandas/core/ops.py]: I think can remove this case if change 492 to
``elif isinstance(ovalues, (datetime.datetime, np.datetime64))``
```
In [4]: np.datetime64('2013-01-01T00:00:00.000000000')
Out[4]: numpy.datetime64('2013-01-01T00:00:00.000000000')

In [5]: isinstance(np.datetime64('2013-01-01T00:00:00.000000000'), np.ndarray)
Out[5]: False

In [6]: isinstance(np.array(np.datetime64('2013-01-01T00:00:00.000000000')), np.ndarray)
Out[6]: True

In [7]: np.array(np.datetime64('2013-01-01T00:00:00.000000000'))
Out[7]: array(1356998400000000000, dtype='datetime64[ns]')
```

File: pandas/core/ops.py (hunk 1 above)
Comment (jreback) [reply pandas/core/ops.py]: ok with removing this code here

File: pandas/core/ops.py (hunk 7 above)
Comment (jreback) [reply pandas/core/ops.py]: I understand, but this is an obscure way of doing this. I would rather actually construct a proper doc-string here

File: pandas/core/ops.py [hunk 12]
```diff
@@ -857,6 +826,8 @@ def wrapper(self, other, axis=None):
         # Validate the axis parameter
         if axis is not None:
             self._get_axis_number(axis)
```
Comment (jreback) [inline pandas/core/ops.py]: yes should change (but agreed may not be hit in tests)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: OK, I'll remove this, can consider docstrings out of scope for this PR.

File: pandas/core/ops.py (hunk 11 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: Sounds good.

**Comment by jreback** [comment]
closing in favor of #19024

**Pull request closed without merging**